### ISSUE TITLE
terraform_remote_state data source requires same version of terraform

### ISSUE THREAD

**Issue by craigedmunds** [hashicorp/terraform#25487]
We have an issue at the moment whereby a 3rd party terraform state file that we use within our project has been produced by an upgraded version of terraform (0.12.26). We currently use terraform 0.12.24. Terraform is validating that these versions align and so currently all our applies are currently failing with the error:

state snapshot was created by Terraform v0.12.26, which is newer than current v0.12.24; upgrade to Terraform v0.12.26 or greater to work with this state

### Terraform Version
We are running 0.12.24 and reading a state file that uses 0.12.26

### Terraform Configuration Files
```hcl
data "terraform_remote_state" "networks" {
  backend = "s3"
  config = {
    bucket = var.source_state_bucket
    region = data.aws_region.current.name
    key    = var.source_networks_tfstate
  }
}
````
### Expected Behavior

We don't want to be forced to upgrade when projects which we depend on upgrade themselves. Perhaps this validation should be optional, maybe by allowing one to specify that the check shouldn't take place:

```hcl
data "terraform_remote_state" "networks" {
  backend = "s3"
  disable_version_check = true
  config = {
    bucket = var.source_state_bucket
    region = data.aws_region.current.name
    key    = var.source_networks_tfstate
  }
}
```

### Actual Behavior

state snapshot was created by Terraform v0.12.26, which is newer than current v0.12.24; upgrade to Terraform v0.12.26 or greater to work with this state

This is also reported in the terraform-provider-aws repo as I wasn't sure where it belongs https://github.com/terraform-providers/terraform-provider-aws/issues/14056

**Comment by PGBI** [comment]
At my company we manage dozens of terraform configs, and we heavily use the `terraform_remote_state` data source to export / import values between these terraform configs.

We would like to start the migration process from terraform 0.12 to 0.13 but we are currently unable to do so because as soon as a terraform config will have been applied using terraform 0.13, its state won't be readable anymore by other terraform configs still on 0.12.

I understand that states are versioned using a separate version number:
```
{
  "version": 4,
  "terraform_version": "0.12.13",
  "serial": 6,
  ...
}
```
Shouldn't `terraform_remote_state` check the `version` property instead of the `terraform_version` one?

**Comment by craigedmunds** [comment]
In lieu of a fix for this @PGBI we switched to using the s3 object provider, and reading the state json with that. It isn’t ideal, but it got us past the problem and it may work for your case too.

**Comment by PGBI** [comment]
> In lieu of a fix for this @PGBI we switched to using the s3 object provider, and reading the state json with that. It isn’t ideal, but it got us past the problem and it may work for your case too.

@craigedmunds Genius!

Some from my testing, the following config...
```
data "terraform_remote_state" "foo" {
  backend = "s3"

  config = {
    bucket   = "my-bucket"
    key      = "foo"
  }
}

locals {
  foo_outputs = data.terraform_remote_state.foo.outputs
}
```

... is strictly equivalent to:
```
data "aws_s3_bucket_object" "foo" {
  bucket = "my-bucket"
  key    = "foo"
}

locals {
  foo_outputs = {for k, v in jsondecode(data.aws_s3_bucket_object.foo.body)["outputs"]: k => v.value }
}
```

Are there any side effects I haven't thought about with this workaround?

Of course this workaround only works for those using the S3 backend.

**Comment by craigedmunds** [comment]
You lose the version checking, which could cause a problem if the state file structure changes, and it seems to require additional IAM permissions on the file - GetObject*

**Comment by alisdair** [comment]
We have now released Terraform versions 0.12.30 and 0.13.6, both of which allow reading remote state from all versions of Terraform at least through 1.0. That is, users of Terraform 0.12.30 and 0.13.6 will be able to access remote state written by other versions of Terraform including 0.11, 0.12, 0.13, and 0.14.

For more details, see the full changelogs:

https://github.com/hashicorp/terraform/blob/v0.12.30/CHANGELOG.md
https://github.com/hashicorp/terraform/blob/v0.13.6/CHANGELOG.md
